libtransmission/lru-cache.h

```diff
@@ -6,44 +6,53 @@
 #pragma once
 
 #include <array>
+#include <condition_variable>
 #include <cstddef> // size_t
 #include <cstdint>
 #include <functional>
+#include <mutex>
+#include <optional>
 #include <utility>
 
+#include "libtransmission/observable.h"
```

Comment:
> Also callers will need to be very careful to never use `auto [entry, tag] = *found;` since that would (I think?) copy the tag and cause a double-call to the destructor lambda?

I don't think this will be a problem, because `ObserverTag`'s copy constructor and copy assignment operator is deleted.

> So I'm wondering if there's some generic RAII wrapper we could use here

That would be a nice thing to have. Are you going to make the PR for it?

P.S. Why was the edit crossed out?

Reply:
> I don't think this will be a problem, because ObserverTag's copy constructor and copy assignment operator is deleted.

:+1:

> That would be a nice thing to have. Are you going to make the PR for it?
>
> P.S. Why was the edit crossed out?

Sorry for not communicating better.

When I investigated scope_guard, I found that they've made a design decision to leave the scope guard's type opaque, so (IIUC) our use of it would be limited to header-based functions that can return `auto`. This would be difficult for returning scoped `tr_sys_file_t` handles, since those implementations happen in per-platform `.cc` files.

I'm going to look at other alternatives and see if any of them are a better fit. I don't think this PR needs to block on that, though -- if you want to move forward with this series, and if you're willing to revisit `ObserverTag` if/when we have some kind of a `scope_guard` type, I'm fine with merging this PR. WDYT?

Reply:
> if you're willing to revisit `ObserverTag` if/when we have some kind of a `scope_guard` type, I'm fine with merging this PR. WDYT?

Sure thing. Go ahead and merge. 👍🏻